### PR TITLE
refactor: replace deprecated `rebind_noauth_module_to_user`, `get_real_user`, `runtime.hostname`, `runtime.course_id` [BD-13]

### PR DESCRIPTION
**Pull request by tecoholic**
## Description 
The `rebind_noauth_module_to_user` function is deprecated<sup>[1]</sup> in the core edx-platform. This is now replaced with a "rebind_user" service. This PR brings this change to the LTI Consumer XBlock.

This should be considered for review and merged after https://github.com/openedx/edx-platform/pull/30320/ is merged.

**Edit**
The PR is now updated with changes related to the deprecation of the `hostname` parameter<sup>[2]</sup> from the runtime as well.

It also replaces usages of `get_real_user` with the `UserService` and `course_id` with `runtime.scope_ids.usage_id.context_key`

This should be considered for merging only after https://github.com/openedx/edx-platform/pull/30308, https://github.com/openedx/edx-platform/pull/30320/, and https://github.com/openedx/edx-platform/pull/30715 are merged.

## Testing instructions

The testing instructions of the parent PR can be used to test the LTI Consumer XBlock.

## Additional information

[1] - https://github.com/openedx/edx-platform/pull/30320/
[2] - https://github.com/openedx/edx-platform/pull/30308

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @tecoholic! I've created [OSPR-6665](https://openedx.atlassian.net/browse/OSPR-6665) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: lti_consumer/lti_xblock.py
Comment (Agrendalath) [inline lti_consumer/lti_xblock.py:145]: It's `rebind_user`.

Otherwise, this will throw `xblock.exceptions.NoSuchServiceError: Service 'rebind_user' was not requested.`.

**Review by Agrendalath — CHANGES_REQUESTED** [review]
I've reused your script from https://github.com/openedx/edx-platform/pull/30320, but it was giving me the following error:
```python
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/views/decorators/clickjacking.py", line 50, in wrapped_view
    resp = view_func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 895, in handle_xblock_callback_noauth
    return _invoke_xblock_handler(request, course_id, usage_id, handler, suffix, course=course)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 1108, in _invoke_xblock_handler
    resp = handler_instance.handle(handler, req, suffix)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/mixins.py", line 84, in handle
    return self.runtime.handle(self, handler_name, request, suffix)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1367, in handle
    return super().handle(block, handler_name, request, suffix=suffix)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/runtime.py", line 1081, in handle
    results = handler(request, suffix)
  File "/edx/shared-src/xblock-lti-consumer/lti_consumer/lti_xblock.py", line 1325, in result_service_handler
    log_authorization_header(request, lti_provider_key, lti_provider_secret)
  File "/edx/shared-src/xblock-lti-consumer/lti_consumer/lti_1p1/oauth.py", line 153, in log_authorization_header
    params = client.get_oauth_params(request)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/oauthlib/oauth1/rfc5849/__init__.py", line 185, in get_oauth_params
    params.append(('oauth_body_hash', base64.b64encode(hashlib.sha1(request.body.decode()).digest()).decode('utf-8')))
TypeError: Unicode-objects must be encoded before hashing
```

To avoid this, I've replaced the following line in `/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/oauthlib/oauth1/rfc5849/__init__.py`.
```diff
- params.append(('oauth_body_hash', base64.b64encode(hashlib.sha1(request.body.encode('utf-8')).digest()).decode('utf-8')))
+ params.append(('oauth_body_hash', base64.b64encode(hashlib.sha1(request.body.decode().encode('utf-8')).digest()).decode('utf-8')))
```

Should I do something different while sending this request?

**Comment by tecoholic** [comment]
@giovannicimolin @Agrendalath I might have jumped the gun on this one. I think I didn't do a proper install inside the container while testing this. Let me rework submit this again for review.

File: lti_consumer/lti_1p1/oauth.py
Comment (tecoholic) [inline lti_consumer/lti_1p1/oauth.py:154]: @Agrendalath Debugging the Oauth failure, I looked at the source for [Webob request](https://github.com/Pylons/webob/blob/259230aa2b8b9cf675c996e157c5cf021c256059/src/webob/request.py#L941) body attribute and the logic in the Oauthlib's `get_oauth_params` [function](https://github.com/oauthlib/oauthlib/blob/master/oauthlib/oauth1/rfc5849/__init__.py#L222). While Webob maintains the content of a request body as binary, oauthlib expects it to be a string.

Both this line here and the oauthlib's code are at least 7 years old and I am not sure how this mismatch didn't raise an issue before. Maybe because, this function was always called with `www-form-urlencoded` data?? 

Anyway, I found that there is alternate function which accpets binary body and can basically do the exact same function in [Line 92](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/lti_1p1/oauth.py#L92). I have reused it here as an replacement and it seems to work without any issues.

Comment (tecoholic) [reply lti_consumer/lti_1p1/oauth.py:154]: Ah.. the CI has caught an error. debugging again

File: lti_consumer/lti_1p1/oauth.py
Comment (tecoholic) [inline lti_consumer/lti_1p1/oauth.py:150]: @Agrendalath Debugging the Oauth failure, I looked at the source for [Webob request](https://github.com/Pylons/webob/blob/259230aa2b8b9cf675c996e157c5cf021c256059/src/webob/request.py#L941) body attribute and the logic in the Oauthlib's `get_oauth_params` [function](https://github.com/oauthlib/oauthlib/blob/master/oauthlib/oauth1/rfc5849/__init__.py#L222). While Webob maintains the content of a request body as binary, oauthlib expects it to be a string.

Both this line here and the oauthlib's code are at least 7 years old and I am not sure how this mismatch didn't raise an issue before. Maybe because, this function was always called with `www-form-urlencoded` data?? 

Anyway, I found that there is alternate function which accpets binary body and can basically do the exact same function in [Line 92](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/lti_1p1/oauth.py#L92). I have reused it here as an replacement.

Since the `oauth_body_hash` is explicitly calculated in the above lines, I think we can consider this a safe replacement to get the other params.

**Comment by tecoholic** [comment]
@giovannicimolin The test I created for testing the LTI 2.0 endpoints in https://github.com/openedx/edx-platform/pull/30320 doesn't seem to create `LtiAgsScore` objects and hence can't trigger the `post_save` [signal](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/signals.py#L17) you have noted. So, I couldn't test it.

Is there a way to test this? Should we setup a LTI 1.3 client for this?

**Comment by giovannicimolin** [comment]
@tecoholic The LTI AGS models are only used by LTI 1.3 with grades enabled.

To avoid the hassle of setting it up, you can:
1. Add an LTI 1.3 block and set the block to scored and add weight.
2. Go to Django admin and create a new `LtiAgsLineItem` and fill in all fields.
3. Still in the Django admin, add a new `LtiAgsScore` and fill in all fields.

This should trigger the signal, you can add a debugger/breakpoint to analyze the parameters and match the conditions. :smile:

File: lti_consumer/lti_1p1/oauth.py
Comment (Agrendalath) [reply lti_consumer/lti_1p1/oauth.py:150]: @tecoholic, good catch! It's working for me now.

However, this is not a part of the runtime usage refactor, so we should move this to a separate commit (or even PR).

**Comment by tecoholic** [comment]
@giovannicimolin I was able to test the the signal is triggered and processed without issue using the following steps:

1. Add a LTI 1.3 Block with **Scored** = `True`, **Weight** default to `1.0` and LTI and I set **Tool launch URL** to `https://saltire.lti.app/tool` so that the student view would load something instead of just showing the "set the url" error.
2. Set the subsection as a graded section (Mid-term Exam)
3. Open the LTI block in LMS after saving and publishing the course
4. An `LtiAgsLineItem` was automatically added when the student view was loaded. Verified by checking the admin (http://localhost:18000/admin/lti_consumer/ltiagslineitem/)
5. Added an external id in the Admin (http://localhost:18000/admin/external_user_ids/externalid/) for the user I was testing, as the signal fetches user using `get_user_from_external_user_id` and the signal was failing without one. I am assuming, normally this gets created when an actual LTI 1.3 tool interacts with the LTI Consumer.
6. Added a `LtiAgsScore` in the admin and used the UUID of the external user id, watched the signal using a debugger.

Everything worked as expected and the score was verified in the gradebook as well.

Does the above testing resolve our concerns?

cc: @Agrendalath

File: lti_consumer/lti_1p1/oauth.py
Comment (tecoholic) [reply lti_consumer/lti_1p1/oauth.py:150]: @Agrendalath I thought about this.
> 
> However, this is not a part of the runtime usage refactor, so we should move this to a separate commit (or even PR).

I think we can move this to a separate commit, but not a separate PR. I tried to see if we can test the rebinding using `www-form-urlencoded` requests and avoid body-hashing, but the rebinding is specifically for [JSON based score updates](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/lti_xblock.py#L1399) and it wouldn't be possible to test it without this tweak/fix.

**Comment by giovannicimolin** [comment]
@tecoholic That's perfect, thank you for testing it!

No more concerns from my side! :+1:

File: lti_consumer/lti_1p1/oauth.py
Comment (Agrendalath) [reply lti_consumer/lti_1p1/oauth.py:150]: Please split this fix into a separate commit then. Leaving everything in a single commit will not give any idea of why we've changed it.

Comment (Agrendalath) [reply lti_consumer/lti_1p1/oauth.py:150]: @tecoholic, ^

**Review by Agrendalath — APPROVED** [review]
@tecoholic, please rebase this on the `master` branch (we need to bump the version to `4.2.0`).

:+1:

* [x]  I tested this: verified that the XBlock is working correctly with the new `rebind_user` service.
* [x]  I read through the code
* [x]  I checked for accessibility issues: n/a
* [x]  Includes documentation: n/a
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

**Comment by Agrendalath** [comment]
Just repeating the note from the description - we shouldn't merge this before https://github.com/openedx/edx-platform/pull/30320.

**Comment by tecoholic** [comment]
@Agrendalath A new commit (05bda984f5647b0559b4605dc8b261d8de59c419) to related to the `hostname` deprecation is also now added as a part of the PR. Kindly review additional changes when you have the time. Thanks in advance.

File: lti_consumer/tests/unit/test_lti_xblock.py
Comment (Agrendalath) [inline lti_consumer/tests/unit/test_lti_xblock.py:1026]: Nit: this is no longer accurate.

Comment (tecoholic) [reply lti_consumer/tests/unit/test_lti_xblock.py]: Fixed this.

**Comment by tecoholic** [comment]
@natabene Thank you for look at this. Looks like https://github.com/openedx/edx-platform/pull/30320 is merged. That means, this should be ready to merge. I am move this to ready.

cc: @Agrendalath

**Comment by Agrendalath** [comment]
@natabene, we would like to merge this after https://github.com/openedx/edx-platform/pull/30715, as this is currently compatible only with the `edx-platform/master` release.

cc: @tecoholic

**Comment by natabene** [comment]
@tecoholic Please coordinate merging with @ormsbee .

**Comment by natabene** [comment]
@tecoholic Is this still work in progress?

**Comment by tecoholic** [comment]
@natabene https://github.com/openedx/edx-platform/pull/30715 is still open. So we can't mark the PR ready until it is merged. @Agrendalath is  coordinating the review for https://github.com/openedx/edx-platform/pull/30715, so we should be move this forward once that's done.

File: lti_consumer/__init__.py
Comment (Agrendalath) [inline lti_consumer/__init__.py:7]: This is a breaking change, so we should bump the major version.

File: lti_consumer/lti_xblock.py
Comment (Agrendalath) [inline lti_consumer/lti_xblock.py:852]: I forgot to mention this, but `self.location` is just [a wrapper](https://github.com/open-craft/edx-platform/blob/4271e24eb90cd2e113eaa2250d807ff67e77a99d/xmodule/modulestore/xml_importer.py#L104-L107) for `self.scope_ids.usage_id`. Could we replace its usage in this XBlock?

File: README.rst
Comment (Agrendalath) [inline README.rst:302]: @xitij2000, it seems that we're maintaining the changelog in the `Changelog.rst` file.

File: lti_consumer/lti_xblock.py
Comment (Agrendalath) [inline lti_consumer/lti_xblock.py:851]: @xitij2000, did you test the whole PR? Because this should break - we have decided not to introduce this service, but rather retrieve this value directly from `settings.LMS_BASE`. I thought we have already reflected it in this PR, but it doesn't seem to be the case.

**Review by Agrendalath — CHANGES_REQUESTED** [review]

File: lti_consumer/lti_xblock.py
Comment (xitij2000) [reply lti_consumer/lti_xblock.py:851]: I did not test the whole PR, just the PR with my changes only to make sure my changes did not break anything. 

I'm sorry I think I misunderstood the scope here. Should I test against this PR and update previous commits as well?

Comment (Agrendalath) [reply lti_consumer/lti_xblock.py:851]: @xitij2000, I now see that it wasn't clear from the ticket's description. Please do.

**Comment by Agrendalath** [comment]
@giovannicimolin, @ashultz0, I'm not familiar with the change introduced in [this commit](https://github.com/openedx/xblock-lti-consumer/pull/294/commits/c14aa1c4bf2009302d81c1a08254f6940fc34dc6), so I just wanted to mention that [this scenario](https://github.com/openedx/xblock-lti-consumer/pull/249#issuecomment-1124419686) (i.e. creating a new `LtiAgsScore` via Django admin) raises the following exception:
```python
2022-11-22 15:20:56,058 ERROR 5002 [django.request] [user None] [ip None] log.py:224 - Internal Server Error: /admin/lti_consumer/ltiagsscore/add/
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/admin/options.py", line 616, in wrapper
    return self.admin_site.admin_view(view)(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/admin/sites.py", line 232, in inner
    return view(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1657, in add_view
    return self.changeform_view(request, None, form_url, extra_context)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/utils/decorators.py", line 43, in _wrapper
    return bound_method(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1540, in changeform_view
    return self._changeform_view(request, object_id, form_url, extra_context)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1586, in _changeform_view
    self.save_model(request, new_object, form, not add)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1099, in save_model
    obj.save()
  File "/edx/shared-src/xblock-lti-consumer/lti_consumer/models.py", line 704, in save
    super().save(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/base.py", line 739, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/base.py", line 787, in save_base
    post_save.send(
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 180, in send
    return [
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 181, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/edx/shared-src/xblock-lti-consumer/lti_consumer/signals/signals.py", line 64, in publish_grade_on_score_update
    raise exc
  File "/edx/shared-src/xblock-lti-consumer/lti_consumer/signals/signals.py", line 53, in publish_grade_on_score_update
    block.set_user_module_score(user, score, block.max_score(), instance.comment)
  File "/edx/shared-src/xblock-lti-consumer/lti_consumer/lti_xblock.py", line 1423, in set_user_module_score
    self.runtime.rebind_noauth_module_to_user(self, user)
  File "/edx/app/edxapp/edx-platform/xmodule/services.py", line 192, in rebind_noauth_module_to_user
    raise RebindUserServiceError(err_msg)
xmodule.services.RebindUserServiceError: rebind_noauth_module_to_user can only be called from a module bound to an anonymous user
```

The current PR doesn't change this behavior, but please let me know if you'd like to review it before we merge.

**Review by Agrendalath — APPROVED** [review]
:+1:

* [x] I tested this: verified that the XBlock is working correctly (with [this exception](https://github.com/openedx/xblock-lti-consumer/pull/249#issuecomment-1323860399)) after replacing deprecated runtime attributes
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation: n/a
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

**Comment by ashultz0** [comment]
@Agrendalath thanks for pointing that out. We'll have to figure out how to let that work without so much magic.

**Comment by openedx-webhooks** [comment]
@tecoholic 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/xblock-lti-consumer&entry.867055334=https://github.com/openedx/xblock-lti-consumer/pull/249&entry.1484655318=https://github.com/tecoholic&entry.752974735=2022-05-10+08:34&entry.1917517419=2022-11-29+10:51&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by ashultz0** [comment]
this is going into the platform today along with a mostly unrelated bugfix as 7.0.1 https://github.com/openedx/edx-platform/pull/31369

File: lti_consumer/outcomes.py
Comment (ashultz0) [inline lti_consumer/outcomes.py:187]: this line turns out to have a bug - it should be        
 real_user = self.xblock.runtime.service(self.xblock, 'user').get_user_by_anonymous_id(anon_id)

**Comment by ashultz0** [comment]
@Agrendalath @tecoholic 7.0.1 failed on edx production in the outcome service and had to be rolled back, I think I have figured out why and noted above

we're actually already changing that to allow external instead of anonymous IDs in 7.0.2 which will I believe fix the problem without having to revert all the rest of the refactors

**Comment by Agrendalath** [comment]
@ashultz0, good catch. Thank you for fixing this!

**Pull request merged**
